### PR TITLE
#9 - python article

### LINKED ISSUE #9: Write blog post about python
At university, I wrote a lot of python scripts to perform some calculations and visualize some data. Usually, I was not allowed to open the repository and was forced to develop/stash everything locally. 
Now, github provides pro accounts for students with the opportunity to create private repos and with the help of Artipie central it's possible to publish python artifacts and stash some data in binary repository privately. 
That's the premise for the blog post. 

I'm planning to create simple python repo with some visualization script, visualization data (large csv/binary file) will be stored in binary repo of Artipie central, python repository will be published (using CD) to Artipie central. After that it should be possible to install python artifact locally, run it and see the plot.

The blog post will describe the whole process and, maybe, I'll manage to come up with some fun university story as a preamble.  

Let's divide the job into several parts:

- [ ] Part one https://github.com/artipie/blog.artipie.com/issues/12

### PR DESCRIPTION
**Pull request by olenagerasimova**
Closes #9 and #12 
I've added python article

### CODE REVIEW

**Comment by olenagerasimova** [comment]
@olegmoz could you please take a look? The article is not finished yet, I want to set up CD in the python example repository and describe configuration here.

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  - python
+  - repository
+  - registry
+  - binary
+---
+
+Being a student I dreamed about something like Artipie, I needed some private centralized storage 
+where I could have kept both binary or text data files and Python packages. We were conducting a research
+and used to work with a lot of data from experimental measurements and used Python to process and 
+visualise them. My university colleagues from that time were mathematicians, they do not have strong
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova it seems that "they do not have strong..." should be "they did not have strong..." as the statement is in past tense

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+development background and were absolutely fine with passing data and code via flash drive or email.
+All my efforts to infiltrate VCS (GitHub, for example, provides free Pro accounts for students
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova "infiltrate" is not right word here, it means to get inside some organization or facility as a spy. I think in such cases it's more common to say about adoption of something by organization, ex "efforts to adopt VCS in university" or "efforts to convince colleagues to adopt VCS"

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+```commandline
+httpie -a https://central.artipie.com/olenagerasimova/data/y1.dat > ./data/y1.dat
+``` 
+
+or simply open [this](https://central.artipie.com/olenagerasimova/data/y1.dat) link in the browser.
+
+To sum up, this is how I'd use Artipie binary storage to store my research data. As this repository 
+API is very simple (http `PUT` and `GET` requests), it's easy to write a peace of code in any language 
+to upload and downloads necessary files. What is also important, authorisation is supported in Artipie, 
+so it'spossible to grant rights to upload/download from the repository how it suits your case.
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova "it'spossible" missing a space between the words

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+To download the file use `GET` request
+
+```commandline
+httpie -a https://central.artipie.com/olenagerasimova/data/y1.dat > ./data/y1.dat
+``` 
+
+or simply open [this](https://central.artipie.com/olenagerasimova/data/y1.dat) link in the browser.
+
+To sum up, this is how I'd use Artipie binary storage to store my research data. As this repository 
+API is very simple (http `PUT` and `GET` requests), it's easy to write a peace of code in any language
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova there is typo, "peace" should be "piece"

**Review by olegmoz — CHANGES_REQUESTED** [review]
@olenagerasimova thanks, the article looks nice. I did some proof-reading, please see the comments. Also I have one suggestion about article's structure. I think it will be better to start with Python project description and later dive into details about file storage. Right now it's not very clear why you need a file storage right after the introduction.

File: _posts/2020/11/2020-11-23-python-repository.md (hunk 2 above)
Comment (olenagerasimova) [reply _posts/2020/11/2020-11-23-python-repository.md]: @olegmoz thanks, I think [inculcate](https://www.lingvolive.com/ru-ru/translate/en-ru/inculcate) will be correct here

**Comment by olenagerasimova** [comment]
@olegmoz thanks, I've corrected mistakes, restructured and finished the article. The only thing missing is the name - do you have any ideas?

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
+---
+layout: post
+date: 2020-11-23
+title: "Some name"
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova I am sure the post needs more meaningful title

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+the main idea of the example is from my university times: download three data files from Artipie Central 
+(the next part is about how these files got there), read the numbers into arrays and 
+use these arrays to draw a plot. Let's use `pip` to install this example package and run it:
+
+```commandline
+python3 -m pip install --index-url https://central.artipie.com/olenagerasimova/pypi/ pypiexample
+python3 -m pypiexample
+``` 
+
+I hope that you see a polar plot with three curves, visualisation of our data files.
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova I'd suggest to rephrase it to "If you run these commands then you will see...", something like that. I was just reading the text and didn't see the plot!

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+``` 
+
+or simply open links [one](https://central.artipie.com/olenagerasimova/data/y1.dat), 
+[two](https://central.artipie.com/olenagerasimova/data/y2.dat) and 
+[three](https://central.artipie.com/olenagerasimova/data/y3.dat) in the browser. 
+The Python example downloads these files and visualise them on the polar plot.
+
+As this binary repository API is very simple (http `PUT` and `GET` requests), it's easy to write 
+a piece of code in any language to upload and downloads necessary files. What is also important, 
+authorisation is supported in Artipie, so it's possible to grant rights to upload/download from
```
Comment (olegmoz) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova I'd suggest to spell "authorization", see https://english.stackexchange.com/questions/282621/authorization-vs-authorisation-im-in-some-real-dilemma

**Review by olegmoz — CHANGES_REQUESTED** [review]
@olenagerasimova thanks for preparing this post, great job. Please see come minor comments

**Comment by olenagerasimova** [comment]
@olegmoz thanks, corrected. Please, check the title - is it okay or maybe you have a better suggestion?

**Comment by olegmoz** [comment]
@olenagerasimova I think it's good

**Comment by olegmoz** [comment]
@yegor256 could you please review this new blog post?

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+quit the postgraduate program and never saw that future to come.
+
+Now, I'd like to show how I could have used Artipie being that student. In the example I'll use 
+[Artipie Central](http://central.artipie.com/), but the same will work with an Artipie instance 
+running on your private server in the private network.
+
+## Python Repository
+
+Artipie supports [PyPi repository](https://github.com/artipie/artipie/tree/master/examples/pypi) which
+`twine` and `pip` can perfectly understand. This means that you can work with Artipie Python
```
Comment (yegor256) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova I would add links to twine and pip. In general, it's good to have links to everything that you mention. This way you help the reader quickly find the right info, if he/she is interested.

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+the main idea of the example is from my university times: download three data files from Artipie Central 
+(the next part is about how these files got there), read the numbers into arrays and 
+use these arrays to draw a plot. Let's use `pip` to install this example package and run it:
+
+```commandline
+python3 -m pip install --index-url https://central.artipie.com/olenagerasimova/pypi/ pypiexample
+python3 -m pypiexample
+``` 
+
+If you run these commands then you will a polar plot with three curves, visualisation of our data files.
```
Comment (yegor256) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova the word "see" is missing?

File: _posts/2020/11/2020-11-23-python-repository.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+use these arrays to draw a plot. Let's use `pip` to install this example package and run it:
+
+```commandline
+python3 -m pip install --index-url https://central.artipie.com/olenagerasimova/pypi/ pypiexample
+python3 -m pypiexample
+``` 
+
+If you run these commands then you will a polar plot with three curves, visualisation of our data files. 
+
+To publish package to the Artipie Central repository build it with `sdist` and use `twine` for upload:
```
Comment (yegor256) [inline _posts/2020/11/2020-11-23-python-repository.md]: @olenagerasimova link to sdist?

**Comment by yegor256** [comment]
@olenagerasimova overall, looks very good. I made a few small comments.

**Comment by olenagerasimova** [comment]
@yegor256 thanks, all fixed

**Comment by yegor256** [comment]
@olegmoz LGTM

**Comment by olegmoz** [comment]
@olenagerasimova @yegor256 thank you, publishing the blog post
